### PR TITLE
Run prettier in deno

### LINKED ISSUE #154: wanted: prettier
currently in deno_std we fork out to Node in order to run prettier:
https://github.com/denoland/deno_std/blob/master/format.ts#L30

This has several problems:
* deno_std shouldn't depend on node.
* different people may have different prettier versions installed
* forking to execute JS is not optimal - we should already be able to execute JS.

Somehow we want to either have a remote link to a version of prettier that can run in the deno (i.e. browser bundle) or (maybe better) drop a prettier bundle into the deno_std source directory with corresponding `.d.ts` file. Let's say `//prettier/prettier.js` and `//prettier/prettier.d.ts`

### PR DESCRIPTION
**Pull request by kt3k**
I copied the standalone version of prettier in `//prettier` and use it from `//format.ts`.

---
Note: I slightly modified `standalone.js`, `parser-markdown.js`, and `parser-typescript.js`. These files use top-level `this` for global object. But deno uses native ES module for loading .js files, and therefore top-level this is undefined. So I changed those top-level `this` occurences to `globalThis` which V8 provides since [v7.1](https://v8.dev/blog/v8-release-71).

closes #154

### CODE REVIEW

File: format.ts [hunk 1]
```diff
@@ -1,36 +1,76 @@
...
+const decoder = new TextDecoder();
+const encoder = new TextEncoder();
+
+function decode(data) {
+  return decoder.decode(data);
+}
+
+function encode(str) {
+  return encoder.encode(str);
+}
```
Comment (ry) [inline format.ts]: I think it's more trouble than it's worth

**Comment by kt3k** [comment]
@ry I found one error for windows (fixed at 15690c7fec36e16e1e45bf1fe85a0205044e803c). Do you use windows now?

File: format.ts (hunk 1 above)
Comment (kt3k) [reply format.ts]: I don't see what exactly you mean. Do you think inlining it like `new TextEncoder().encode(...)` is better?

Comment (ry) [reply format.ts]: inlining `encoder.encode(str)` isn't much more to write than `encode(str)` - and it prevents two extra functions

File: format.ts [hunk 2]
```diff
@@ -1,36 +1,136 @@
...
+): Promise<void> {
+  const text = new TextDecoder().decode(await readFile(filename));
+  const formatted = prettier.format(text, {
+    parser,
+    plugins: prettierPlugins
+  });
+
+  if (text !== formatted) {
+    console.log(`Formatting ${filename}`);
+    await writeFile(filename, new TextEncoder().encode(formatted));
```
Comment (ry) [inline format.ts]: Having a file-local encoder/decoder is fine - maybe better than creating a new one each time?

**Comment by ry** [comment]
Works for me now.. not sure what was wrong before. I'm happy with this as soon as CI is green.

File: format.ts [hunk 3]
```diff
@@ -1,36 +1,136 @@
...
+  parser: "typescript" | "markdown"
+): Promise<boolean> {
+  const text = new TextDecoder().decode(await readFile(filename));
+  const formatted = prettier.check(text, {
+    parser,
+    plugins: prettierPlugins
+  });
+
+  if (!formatted) {
+    console.log(`${filename} ... Not formatted`);
```
Comment (hayd) [inline format.ts]: console.error ?

Is it possible to print anything about the formatting issue?

Comment (kt3k) [reply format.ts]: I changed it to console.error

> Is it possible to print anything about the formatting issue?

I agree that we should print some more details, but I don't find any text diff library usable in deno. So I put TODO comment here for now.

**Comment by kt3k** [comment]
@hayd 
> Should this be exposed so other projects can use it?

I added the entrypoint of prettier scripts. It should be usable like the below after being merged to master.

```
import { prettier, prettierPlugins } from "https://deno.land/x/std/prettier/prettier.ts";

prettier.format("const x = 1", { plugins: prettierPlugins }); // => "const x = 1;"
```

**Comment by kt3k** [comment]
Azure pipeline seems merging the branch to master temporarily before running CI process, and that caused the CI failure above. (master branch contained some unformatted code).

**Comment by hayd** [comment]
I mean could you be able to call `deno deno.land/x/prettier/main.ts` (or whatever) and if would format the directory you're in. I.e. a good default for any deno repo to use.

File: format.ts [hunk 4]
```diff
@@ -1,36 +1,135 @@
...
-    exit(s.code);
+}
+
+// Gets the source files in the repository
+async function getSourceFiles() {
+  return decoder
+    .decode(
+      await readAll(
+        xrun({
+          args: ["git", "ls-files"],
```
Comment (hayd) [inline format.ts]: One thing to note: This includes deleted files. I don't think it's quite right, but I haven't been able to work out the correct modification...

Comment (kt3k) [reply format.ts]: Thanks for pointing! I'll try fix this in another PR.

Comment (hayd) [reply format.ts]: Perhaps there ought to be recursive and name only options for readDir in deno!

Comment (hayd) [reply format.ts]: Other than this issue, I think this format file could be used as is by external projects.

**Pull request merged**